src/core/geometry/qgsgeometryutils.cpp

```diff
@@ -1257,3 +1265,533 @@ bool QgsGeometryUtils::transferFirstZValueToPoint( const QgsPointSequence &point
 
   return rc;
 }
+
+QgsPoint QgsGeometryUtils::createPointWithMatchingDimensions( double x, double y, const QgsPoint &reference )
+{
+  if ( reference.is3D() && reference.isMeasure() )
+    return QgsPoint( Qgis::WkbType::PointZM, x, y, 0.0, 0.0 );
+  else if ( reference.is3D() )
+    return QgsPoint( Qgis::WkbType::PointZ, x, y, 0.0 );
+  else if ( reference.isMeasure() )
+    return QgsPoint( Qgis::WkbType::PointM, x, y, 0.0, 0.0 );
+  else
+    return QgsPoint( x, y );
+}
+
+QgsPoint QgsGeometryUtils::interpolatePointOnSegment( double x, double y,
+    const QgsPoint &segStart, const QgsPoint &segEnd,
+    double distanceFromStart )
+{
+  QgsPoint result = createPointWithMatchingDimensions( x, y, segStart );
+
+  if ( segStart.is3D() && segEnd.is3D() && result.is3D() )
+  {
+    double z1 = segStart.z();
+    double z2 = segEnd.z();
+    double interpolatedZ;
+    double tempX, tempY;
+    QgsGeometryUtilsBase::pointOnLineWithDistance(
+      segStart.x(), segStart.y(), segEnd.x(), segEnd.y(),
+      distanceFromStart, tempX, tempY, &z1, &z2, &interpolatedZ );
+    result.setZ( interpolatedZ );
+  }
+
+  if ( segStart.isMeasure() && segEnd.isMeasure() && result.isMeasure() )
+  {
+    double m1 = segStart.m();
+    double m2 = segEnd.m();
+    double interpolatedM;
+    double tempX, tempY;
+    QgsGeometryUtilsBase::pointOnLineWithDistance(
+      segStart.x(), segStart.y(), segEnd.x(), segEnd.y(),
+      distanceFromStart, tempX, tempY, nullptr, nullptr, nullptr, &m1, &m2, &interpolatedM );
+    result.setM( interpolatedM );
+  }
+
+  return result;
+}
+
+bool QgsGeometryUtils::createChamfer( const QgsPoint &seg1Start, const QgsPoint &seg1End,
+                                      const QgsPoint &seg2Start, const QgsPoint &seg2End,
+                                      double distance1, double distance2,
+                                      QgsPoint &chamferStart, QgsPoint &chamferEnd,
+                                      double epsilon )
+{
+  // Apply symmetric distance if distance2 is negative
+  if ( distance2 < 0 )
```

Comment:
```suggestion
  if ( distance2 <= 0 )
```
i don't think there's any use case for a 0 distance?